### ISSUE TITLE
Multiple Repositories Hosts Servers

### ISSUE THREAD

**Issue by scapy11** [pgbackrest/pgbackrest#1443]
Please provide the following information when submitting an issue (feature requests or general comments can skip this):

1. pgBackRest version: 2.34

2. PostgreSQL version: 13

3. Operating system/version - if you have more than one server (for example, a database server, a repository host server, one or more standbys), please specify each:  RHEL8

4. Did you install pgBackRest from source or from a package? Package

7. Describe the issue:
Hi, 

Don't know if is possible, but I want to create 2 repo server. This is possible?
One will be the central (repo1) and if for some reason it have any downtime to be able to make backups from another backup server (repo2).

Already tried to create the stanza on repo 2 server, but give me error: `ERROR: [027]: pg and repo hosts cannot both be configured as remote`

This is the pgbackrest.conf on bck-server1:

```
[global]
repo1-cipher-pass=xxxx
repo1-cipher-type=aes-256-cbc
repo1-path=/pgbackrest/backups
repo1-retention-full=24
repo1-retention-full-type=time
repo2-cipher-pass=xxx
repo2-cipher-type=aes-256-cbc
repo2-host=bck-server2
repo2-host-user=pgbackrest
repo2-path=/pgbackrest/backups
repo2-retention-full=24
repo2-retention-full-type=time

archive-async=y
spool-path=/pgbackrest/log
log-level-console=info
log-level-file=info
process-max=2

[demo]
repo1-path=/pgbackrest
repo2-path=/pgbackrest
pg1-host=pg-server
pg1-path=/pgdata/pg13
pg1-port=5432
pg1-host-user=postgres
```

On bck-server2:

```
[global]
repo1-cipher-pass=xxxxx
repo1-cipher-type=aes-256-cbc
repo1-host=bck-servee2
repo1-host-user=pgbackrest

repo2-cipher-pass=xxx
repo2-cipher-type=aes-256-cbc
repo2-path=/pgbackrest/backups
repo2-retention-full=24
repo2-retention-full-type=time
archive-async=y
spool-path=/pgbackrest/log
log-level-console=info
log-level-file=info
process-max=2

[demo]
repo1-path=/pgbackrest
repo2-path=/pgbackrest
pg1-host=pg-server
pg1-path=/pgdata/pg13
pg1-port=5432
pg1-host-user=postgres
```

This can be done or only one backup server can be possible ?

Thanks

**Comment by pgstef** [comment]
Hi,

First of all, the error you get is because `repo-host` and `pg-host` can't be used together. Meaning you can't pull a backup from a remote db and push it to a remote backup server.

What you'd have to do is configure your `repo1` + `pg1` stanza on your 2 backup servers (they don't need to know each other in the config) and then, on your db host, specify both `repo1-host` and `repo2-host` so the `archive-push` command can push WAL archives to both locations. However, the archives pushed on your `repo2` won't be expired until a backup is taken in that repository.

The archiving process happens on the db host, so the `archive-async=y` and `spool-path` won't be used on your backup servers.

Also, I find some of your configuration a bit weird:
* `spool-path=/pgbackrest/log`: spool-path is used for archives, not logs
* `repo1-path=/pgbackrest/backups` overcharged by `repo1-path=/pgbackrest`

At the end, this could be a working example I guess:

* Same config on **bck-server1** and **bck-server2**

```ini
[global]
repo1-path=/pgbackrest/backups
repo1-cipher-pass=xxxx
repo1-cipher-type=aes-256-cbc
repo1-retention-full=24
repo1-retention-full-type=time

log-level-console=info
log-level-file=debug
start-fast=y
delta=y
process-max=2

[demo]
pg1-host=pg-server
pg1-host-user=postgres
pg1-path=/pgdata/pg13
```

* On **pg-server**

```ini
[global]
repo1-host=bck-server1
repo1-host-user=pgbackrest
repo1-host=bck-server2
repo1-host-user=pgbackrest

archive-async=y
spool-path=/pgbackrest/spool
log-level-console=info
log-level-file=debug
delta=y
process-max=2

[demo]
pg1-path=/pgdata/pg13
pg1-port=5432			# default value
pg1-user=postgres		# default value
```

Of course, you can adjust the number of processes used with `process-max` according to your servers specs.

Hope this helps,
Kind regards

**Comment by scapy11** [comment]
Thanks @pgstef  for your reply. Yesterday tried to config like you said but give me always this error if I  add repo2 on pg-server config. Because of that I came here to know if this is a possible solution.

Error:
```
WARN: unable to check pg-1: [JsonFormatError] expected '{' at 'logout'
ERROR: [027]: no database found
       HINT: check indexed pg-path/pg-host configurations
2021-06-30 10:44:09.061 P00   INFO: check command end: aborted with exception [027]
```

Regarding with the parameters you mention before:

- `spool-path` - normally I put it on pg_wal directory, but for now I save it to log file only for this PoC.
- `repo1-path` - I config a repo1-path for global if I don't have setting on stanza. In this case the demo stanza will save the backup always to /pgbackrest directory

Thanks for you support

**Comment by pgstef** [comment]
That's a strange error. You should probably check your pgbackrest version on the db host, and validate the configuration files content. I tested the config I wrote earlier and it's working.

**Comment by scapy11** [comment]
All servers have the same pgbackrest version
```
$ pgbackrest version
pgBackRest 2.34
```

For now I only can perform backup on bck-server1, if on pg-server I add repo2 parameters the check command failed.
This is my configuration:

- on bck-server1 and bck-server2 (the cipher-pass is the same on both)
```
[global]
repo1-cipher-pass=xxxxx
repo1-cipher-type=aes-256-cbc
repo1-path=/pgbackrest/backups
repo1-retention-full=24
repo1-retention-full-type=time

log-level-console=info
log-level-file=info
process-max=2

[demo]
repo1-path=/pgbackurest
pg1-host=pg-server
pg1-path=/pgdata/pg13
pg1-port=5432
pg1-host-user=postgres
```

- on pg-server
```
repo1-cipher-pass=xxx
repo1-cipher-type=aes-256-cbc
repo1-host=bck-server1
repo1-host-user=pgbackrest
#repo2-cipher-pass=xxxx
#repo2-cipher-type=aes-256-cbc
#repo2-host=bck-server2
#repo2-host-user=pgbackrest
log-level-console=info
log-level-file=info
archive-async=y
spool-path=/pgbackrest/log

[demo]
repo1-path=/pgbackuprest
#repo2-path=/pgbackuprest
pg1-path=/pgdata/pg13
pg1-user=postgres
pg1-port=5432
```

**Comment by pgstef** [comment]
That doesn't look like what I gave you...

On **bck-server1** and **bck-server2**:
* `repo1-path=/pgbackurest` -> typo ?
* `pg1-port=5432` -> useless here

On **pg-server** don't provide `repo1-path` nor `repo2-path`, those should be configured on the backup servers, not here.
Not sure about the cipher-pass though, it should probably work without it here too.

**Comment by scapy11** [comment]
if I don't provide `repo1-path` on pg-server it give me error:

```
ERROR: [103]: unable to find a valid repository:
       repo1: [FileOpenError] unable to load info file '/var/lib/pgbackrest/archive/pg-server/archive.info' or '/var/lib/pgbackrest/archive/pg-server/archive.info.copy':
       FileOpenError: raised from remote-0 protocol on 'bck-server1': unable to open file '/var/lib/pgbackrest/archive/pg-server/archive.info' for read: [13] Permission denied
       FileOpenError: raised from remote-0 protocol on 'bck-server1': unable to open file '/var/lib/pgbackrest/archive/pg-server/archive.info.copy' for read: [13] Permission denied
```

And if I don't provide the cipher pass on pg-sever config, it gave the error:
```
 FormatError: key/value found outside of section at line 1: Salted__▒▒▒▒▒A`
```

thanks

**Comment by pgstef** [comment]
What command gives you what error exactly ? 
I confirm, don't define `repo1-path` on the db host or it will try to act locally. (complete working example [here](https://pgstef.github.io/2020/11/21/combining_pgbackrest_dedicated_repository_host_and_streaming_replication.html))

Are you sure your ssh connection works ? Are you sure your dedicated backup user on the backup server has enough perms on the `repo1-path` there ?

`/var/lib/pgbackrest/archive/pg-server/archive.info` is also a very weird path. It would be the default `repo1-path` on **bck-server1** (so no definition for that in the global section) and using `--stanza=pg-server` where you should use `--stanza=demo`.

So, I'm sorry but the errors you're describing don't match the configuration you provide and only extracts of what you really have won't help here. Given that the example I gave you above is working (when ssh access and directory permissions are set correctly), I suggest you clean up your setup / configuration / commands and test again.

**Comment by scapy11** [comment]
the command I used is:
` pgbackrest --stanza=demo check`

I change the names of stanza and servers for security. because of that appear pg-server instead of demo (already updated).
yes, don't know why it pick always /var/lib/.. when I don't provide the `repo1-path`, it should do the ssh on bck-server1 and pick the config from there

Already did a clean up and setup from scratch even without cipher and the same issue happens.

Thanks for your support

**Comment by pgstef** [comment]
It is connecting to **bck-server1**: `raised from remote-0 protocol on 'bck-server1'`. If it uses the default value, it probably can't locate the config.

Where are you configuration files located ? Perhaps try adding `--repo1-host-config` to check if the issue comes from there

**Comment by scapy11** [comment]
I forgot to add the `repo1-host-config-path` and now don't raise any error.
But thanks for your tips, I already read in the past your blog and you are a good contributor for this amazing tool.
So, I tried to config the bck-server2 as yesterday I figured out, but forgot to do. Now I'm able to have 2 backups servers to manage all backup for any cluster pg database. On bck-server2 to be able to do the backups and receive the wal's I had to config as repo2 instead of repo1, because is always a mismatch on pg-server. 

So here is the config, if someone pretend to use for repository host redundancy, if for example repo1 host is down for maintenance or for a crash, can always use a second repo host to manage the backups:

- on **bck-server1**
```
[global]
repo1-path=/pgbackrest/backups
repo1-retention-full=24
repo1-retention-full-type=time

log-level-console=info
log-level-file=info
process-max=2

[demo]
repo1-path=/pgbackrest
pg1-host=pg-server
pg1-path=/pgdata/pg13
```

- on **bck-server2**
```
[global]
repo2-path=/pgbackrest/backups
repo2-retention-full=24
repo2-retention-full-type=time

log-level-console=info
log-level-file=info
process-max=2

[demo]
repo2-path=/pgbackrest
pg1-host=pg-server
pg1-path=/pgdata/pg13
```

- on **pg-server**
```
[global]
repo1-host=bck-server1
repo1-host-user=pgbackrest
repo1-host-config-path=/pgbackrest/conf
repo2-host=bck-server2
repo2-host-user=pgbackrest
repo2-host-config-path=/pgbackrest/conf
log-level-console=info
log-level-file=info
archive-async=y
spool-path=/pgbackrest/log

[demo]
pg1-path=/pgdata/pg13
```

I will recreate this again to encrypt the repos.

Thanks a lot for your support @pgstef 

Best Regards

**Comment by pgstef** [comment]
Hm, you're right, it will ask for `repo2` on **bck-server2**. Not sure why it worked with repo1 in my test.
Anyway, glad it works for you now :-)

Just pay attention that the `archive_command = pgbackrest archive-push....` will try to push archives to all defined repositories. So if 1 of your 2 backup servers, PG will keep its wal files until it comes back online. Given you use the async archiving, you'll get the fault tolerance and the archives will continue to come in the available repo even if the other is down.

Since archives are expired from the repo based on backups, you will probably have to clean-up manually the archives from the repo you don't use for backups...

Kind regards

**Comment by scapy11** [comment]
yes, because of that I use async for archives on clusters.
A tricky thing is when I want to delete the stanza on repo2, don't know why I need to change on pgbackrest.conf on bck-server2 the `repo2-path` on demo stanza to `repo1-path`, if not it raise the error. I think is because it pick from pg-server the repo1-path:
```
pgbackrest --stanza=demo stanza-delete --repo=2
2021-06-30 15:16:35.482 P00   INFO: stanza-delete command begin 2.34: --config-path=/pgbackrest/conf --exec-id=2225239-a5c89e1e --log-level-console=info --log-level-file=info --log-path=/pgbackrest/log --pg1-host=pg-server --pg1-host-user=postgres --pg1-path=/pgdata/pg13 --pg1-port=5432 --repo=2 --repo2-cipher-pass=<redacted> --repo2-cipher-type=aes-256-cbc --repo1-path=/pgbackrest --repo2-path=/pgbackrest/backups --stanza=edbt5b2x
ERROR: [099]: expected '{' at 'logout'
2021-06-30 15:16:35.702 P00   INFO: stanza-delete command end: aborted with exception [099]
```

I maybe will sometime run a backup from bck-server2, but I will put a script to cleanup the archives.
My idea is after backups to both repos, every night run a backup to tape too and expire some backups from disks to save space.
If need it, restore from tape to any bck-server and run a restore on any pg-server need it.

Kind Regards

**Comment by pgstef** [comment]
I don't think it goes to db host. I believe it's because you have `repo1-path=/pgbackrest` configured in the global part. That's also one reason why I don't like to have repo-paths configured in the stanza section.

The error you get is kinda strange though, I never saw it before

**Comment by dwsteele** [comment]
> ERROR: [099]: expected '{' at 'logout'

pgBackRest does not use the word `logout` in its protocol, so this looks like some noise from your SSH setup, i.e. on exit your SSH server is printing 'logout'.

**Comment by scapy11** [comment]
SSH is fine, but every time it pgbackrest check something wrong on config file, it abort with that error.

**Comment by dwsteele** [comment]
> SSH is fine, but every time it pgbackrest check something wrong on config file, it abort with that error.

It doesn't look that way to me. I can't see how the `logout` is coming from us. But an error result from pgbackrest might be causing some other issue.

**Comment by scapy11** [comment]
the `logout`  message is from ssh connection it leave the target host server.

**Comment by scapy11** [comment]
ok, already found it, someone put a trap on profile. So the full error is:
```
pgbackrest --stanza=edbt5b2x stanza-delete --repo=2
2021-06-30 16:27:20.153 P00   INFO: stanza-delete command begin 2.34: --config-path=/pgbackrest/conf --exec-id=2235986-10affa8e --log-level-console=info --log-level-file=info --log-path=/pgbackrest/log --pg1-host=pg-server --pg1-host-user=postgres --pg1-path=/pgdata/pg13 --pg1-port=5432 --repo=2 --repo2-cipher-pass=<redacted> --repo2-cipher-type=aes-256-cbc --repo2-path=/pgbackrest --stanza=demo
ERROR: [032]: remote-0 process on 'pg-server' terminated unexpectedly [32]: ERROR: [032]: key '2' is not valid for 'repo' option
2021-06-30 16:27:20.340 P00   INFO: stanza-delete command end: aborted with exception [032]
```

**Comment by pgstef** [comment]
I managed to reproduce the error with a pretty simple configuration: 

```ini
# bck-server1
[global]
repo1-path=/backup_space
repo1-retention-full=1
log-level-console=info
log-level-file=debug
start-fast=y

[demo]
pg1-host=pg-server
pg1-path=/var/lib/pgsql/13/data

# bck-server2
[global]
repo2-path=/backup_space
repo2-retention-full=1
log-level-console=info
log-level-file=debug
start-fast=y

[demo]
pg1-host=pg-server
pg1-path=/var/lib/pgsql/13/data

# pg-server
[global]
repo1-host=bck-server1
repo1-host-user=pgbackrest
repo2-host=bck-server2
repo2-host-user=pgbackrest

archive-async=y
log-level-console=info
log-level-file=debug
delta=y
process-max=2

[demo]
pg1-path=/var/lib/pgsql/13/data
```

Trying to remove the stanza from the backup server fails:

```bash
bck-server2$ pgbackrest --stanza=demo stanza-delete --repo=2
INFO: stanza-delete command begin 2.34: --exec-id=3869-be0d3768 --log-level-console=info --log-level-file=debug --pg1-host=pg-server --pg1-path=/var/lib/pgsql/13/data --repo=2 --repo2-path=/backup_space --stanza=demo
ERROR: [032]: remote-0 process on 'pg-server' terminated unexpectedly [32]: ERROR: [032]: key '2' is not valid for 'repo' option
INFO: stanza-delete command end: aborted with exception [032]
```

Succeed from the db server:

```bash
pg-server$ pgbackrest info
stanza: demo
    status: mixed
        repo1: ok
        repo2: error (no valid backups)
    cipher: none

pg-server$ pgbackrest --stanza=demo stanza-delete --repo=2
INFO: stanza-delete command begin 2.34: --exec-id=4228-8def1dc6 --log-level-console=info --log-level-file=debug --pg1-path=/var/lib/pgsql/13/data --repo=2 --repo1-host=bck-server1 --repo2-host=bck-server2 --repo1-host-user=pgbackrest --repo2-host-user=pgbackrest --stanza=demo
INFO: stanza-delete command end: completed successfully (462ms)
```

By habit, since all the repos are configured on the db host, I tend to execute the `stanza-*` (and `check`) commands from the db host directly. Since it works on **bck-server1**, I don't exactly know why it fails from **bck-server2** though..

**Comment by scapy11** [comment]
I don't know too why this happen. But if you change on **bck-server2** the stanza section to `repo1-path` instead of `repo2-path`, you will be able to delete from repo server.

**Comment by dwsteele** [comment]
This should be fixed by #1480. Let us know if you continue to see the issue in 2.35.
